File: frontend/src/api/public.ts
// Deliberately does NOT use apiFetch (api/client.ts) - this is called from
// SharedProjectScreen, which has no API key and must not touch the authenticated
// client's 401/403 handling. See docs/agent/conventions.md § Invariants.

export interface PublicProjectItem {
  name: string;
  quantity: number;
  quantity_completed: number;
}

export interface PublicProjectPart {
  name: string;
  quantity: number;
}

export interface PublicProjectLink {
  url: string;
  label: string | null;
}

export interface PublicProject {
  name: string;
  customer: string;
  due_date: string | null;
  on_hold: boolean;
  items: PublicProjectItem[];
  parts: PublicProjectPart[];
  links: PublicProjectLink[];
  jobs_total: number;
  jobs_complete: number;
  estimate_seconds_remaining: number | null;
  updated_at: string;
}

export async function getPublicProject(token: string): Promise<PublicProject> {
  const resp = await fetch(`/api/v1/public/projects/${encodeURIComponent(token)}`);
  if (!resp.ok) {
    throw new Error(resp.status === 404 ? 'not-found' : `${resp.status}`);
  }
  return resp.json();
}

File: backend/app/api/routes/public.py
"""Public, unauthenticated read-only routes.

This is the ONE deliberate exception (beyond the empty-api_keys-table bootstrap
hatch) to this codebase's "every /api/v1/* route requires Depends(require_scope(...))"
invariant - see docs/agent/conventions.md § Invariants. Keep this module small and
its own file so the one auth-exempt surface in the whole API stays trivially easy to
audit in isolation. Never import require_scope here, and never add a route to this
router that isn't addressed by an unguessable per-resource token.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ...database import get_session
from ...models import Job, Project
from .projects import _load_items, _load_links, _load_parts, _project_progress

router = APIRouter(prefix="/api/v1/public", tags=["public"])


@router.get("/projects/{token}", summary="Get a project via its public share token")
async def get_public_project(token: str, session: AsyncSession = Depends(get_session)) -> dict:
    proj = (await session.execute(
        select(Project).where(Project.share_token == token)
    )).scalar_one_or_none()
    if proj is None:
        raise HTTPException(404, "Not found")

    items = await _load_items(session, proj.id)
    parts = await _load_parts(session, proj.id)
    links = await _load_links(session, proj.id)
    job_rows = (await session.execute(
        select(Job).where(Job.project_id == proj.id)
    )).scalars().all()
    progress = _project_progress(job_rows)

    return {
        "name": proj.name,
        "customer": proj.customer,
        "due_date": proj.due_date,
        "on_hold": proj.on_hold,
        "items": [
            {"name": i["file_name"], "quantity": i["quantity"], "quantity_completed": i["quantity_completed"]}
            for i in items
        ],
        "parts": [{"name": p["name"], "quantity": p["quantity"]} for p in parts],
        "links": [{"url": l["url"], "label": l["label"]} for l in links],
        "jobs_total": progress["jobs_total"],
        "jobs_complete": progress["jobs_complete"],
        "estimate_seconds_remaining": progress["estimate_seconds_remaining"],
        "updated_at": proj.updated_at,
    }
